### PR TITLE
Serial timeout

### PR DESCRIPTION
**Pull request by vortex314**
Added serial timeout and freertos to espidf.
Succesfull in establishing serial sessions continuously

### CODE REVIEW

**Comment by vortex314** [comment]
image

**Comment by vortex314** [comment]
Peek 2024-03-22 23-02

File: src/system/espidf/system.c [hunk 1]
```diff
@@ -50,48 +50,77 @@
...
-        z_arg->_arg = arg;
-        if (xTaskCreate((void *)z_task_wrapper, "", 5120, z_arg, configMAX_PRIORITIES / 2, task) != pdPASS) {
-            ret = -1;
+    if (z_arg == NULL) {
+        return -1;
+    }
+
+    z_arg->fun = fun;
+    z_arg->arg = arg;
+    z_arg->join_event = task->join_event = xEventGroupCreate();
```
[the remark this answers is not in the document]
Comment (jean-roland) [reply src/system/espidf/system.c:85]: Could you split the double assignment in two separate lines?

[the remark this answers is not in the document]
Comment (vortex314) [reply src/system/espidf/system.c:85]: copied this from freertos_plus_tcp/system.c line 90, looks like MISRA is a recent requirement ?

File: src/system/espidf/system.c [hunk 2]
```diff
@@ -50,48 +50,77 @@
...
+
+    if (attr == NULL) {
+        attr = &z_default_task_attr;
```
[the remark this answers is not in the document]
Comment (vortex314) [reply src/system/espidf/system.c:88]: Is there a way to validate MISRA locally , before committing ?

File: src/system/espidf/system.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jean-roland) [reply src/system/espidf/system.c:85]: It's not recent but the workflow wasn't activated on your previous PR, also we ignore some of them, like 15.5.

File: src/system/espidf/system.c [hunk 3]
```diff
@@ -50,48 +50,77 @@
...
 int8_t z_task_join(z_task_t *task) {
-    // Note: task/thread join not supported on FreeRTOS API, so we force its deletion instead.
-    return zp_task_cancel(task);
+    xEventGroupWaitBits(task->join_event, 1, pdFALSE, pdFALSE, portMAX_DELAY);
+    return 0;
 }
 
-int8_t zp_task_cancel(z_task_t *task) {
-    vTaskDelete(*task);
+int8_t z_task_cancel(z_task_t *task) {
```
[the remark this answers is not in the document]
Comment (vortex314) [reply src/system/espidf/system.c:116]: This should be straightforward @jean-roland ? I have no clue.

[the remark this answers is not in the document]
Comment (vortex314) [reply src/system/espidf/system.c:116]: Is there a misra handbook ?

[the remark this answers is not in the document]
Comment (vortex314) [reply src/system/espidf/system.c:116]: https://www.law.cornell.edu/rules/fre/rule_804 ?  ;-)

[the remark this answers is not in the document]
Comment (jean-roland) [reply src/system/espidf/system.c:116]: No this one we have no clue too, just the two I mentioned.

File: src/system/espidf/system.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mallets) [reply src/system/espidf/system.c:88]: That requires a commercial license, so I'm afraid it's not easy to validate but on the GitHub CI.

**Comment by jean-roland** [comment]
Alright, well seeing that there's only small details that we're more used to handle, I suggest we merge your PR as is and I'll do a small fix PR for the MISRA stuff.

**Review by jean-roland — APPROVED** [review]

**Comment by Mallets** [comment]
Merging into `serial_timeout` branch first. 
We take care os resolving the conflict and to merge it into main.

**Comment by vortex314** [comment]
Thanks for all the help guys  !

**Comment by sashacmc** [comment]
Closes: https://github.com/eclipse-zenoh/zenoh-pico/issues/408

**Pull request merged**